infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/server/KubernetesServerExposer.java

```diff
@@ -178,11 +184,43 @@ public void expose(Map<String, ? extends ServerConfig> servers) throws Infrastru
     splitServersAndPortsByExposureType(
         servers, internalServers, externalServers, secureServers, unsecuredPorts, securedPorts);
 
+    provisionServicesForDiscoverableServers(servers);
+
     exposeNonSecureServers(internalServers, externalServers, unsecuredPorts);
 
     exposeSecureServers(secureServers, securedPorts);
   }
 
+  // TODO: this creates discoverable services as an extra services. Service for same {@link
+  //  ServerConfig} is also created later in in {@link #exposeNonSecureServers(Map, Map, Map)} or
+  //  {@link #exposeSecureServers(Map, Map)} as a non-discoverable one. This was added during
+  //  working on adding endpoints for kubernetes/openshift components, to keep behavior consistent.
+  //  However, this logic is probably broken and should be changed.
+  /**
+   * Creates services with defined names for discoverable {@link ServerConfig}s. The name is taken
+   * from {@link ServerConfig}'s attributes under {@link ServerConfig#SERVER_NAME_ATTRIBUTE} and
+   * must be set, otherwise service won't be created.
+   */
+  private void provisionServicesForDiscoverableServers(
+      Map<String, ? extends ServerConfig> servers) {
+    servers.forEach(
+        (k, server) -> {
```

Comment:
wdyt may be `classical` loop may look simpler?

Reply:
```java
    for (String serverName : servers.keySet()) {
      ServerConfig server = servers.get(serverName);
      if (server.isDiscoverable()
          && server.getAttributes().containsKey(SERVER_NAME_ATTRIBUTE)) {
        ServicePort servicePort = getServicePort(server);
        Service service =
            new ServerServiceBuilder()
                .withName(server.getAttributes().get(SERVER_NAME_ATTRIBUTE))
                .withMachineName(machineName)
                .withSelectorEntry(CHE_ORIGINAL_NAME_LABEL, pod.getMetadata().getName())
                .withPorts(Collections.singletonList(servicePort))
                .withServers(Collections.singletonMap(serverName, server))
                .build();
        k8sEnv.getServices().put(service.getMetadata().getName(), service);
      }
    }
```

or 

```java
    for (Entry<String, ? extends ServerConfig> serverEntry : servers.entrySet()) {
      String serverName = serverEntry.getKey();
      ServerConfig server = serverEntry.getValue();
      if (server.isDiscoverable()
          && server.getAttributes().containsKey(SERVER_NAME_ATTRIBUTE)) {
        ServicePort servicePort = getServicePort(server);
        Service service =
            new ServerServiceBuilder()
                .withName(server.getAttributes().get(SERVER_NAME_ATTRIBUTE))
                .withMachineName(machineName)
                .withSelectorEntry(CHE_ORIGINAL_NAME_LABEL, pod.getMetadata().getName())
                .withPorts(Collections.singletonList(servicePort))
                .withServers(Collections.singletonMap(serverName, server))
                .build();
        k8sEnv.getServices().put(service.getMetadata().getName(), service);
      }
    }
```

I don't know, looks same for me. I don't have strong preference here.

Reply:
from the other end, here's the stream version
```java
    servers
        .entrySet()
        .stream()
        .filter(
            e ->
                e.getValue().isDiscoverable()
                    && e.getValue().getAttributes().containsKey(SERVER_NAME_ATTRIBUTE))
        .map(
            e ->
                new ServerServiceBuilder()
                    .withName(e.getValue().getAttributes().get(SERVER_NAME_ATTRIBUTE))
                    .withMachineName(machineName)
                    .withSelectorEntry(CHE_ORIGINAL_NAME_LABEL, pod.getMetadata().getName())
                    .withPorts(Collections.singletonList(getServicePort(e.getValue())))
                    .withServers(Collections.singletonMap(e.getKey(), e.getValue()))
                    .build())
        .forEach(service -> k8sEnv.getServices().put(service.getMetadata().getName(), service));
```

Reply:
updated to for loop